packages/transactions/src/clarity/prettyPrint.ts

```diff
@@ -0,0 +1,138 @@
+ * @exemple
+ * ```ts
+ * formatList(Cl.list([Cl.uint(1)]))
+ * // (list u1)
+ *
+ * formatList(Cl.list([Cl.uint(1)]), 2)
+ * // (list
+ * //   u1
+ * // )
+ * ```
+ */
+function formatList(cv: ListCV, space: number, depth = 1): string {
+  if (cv.list.length === 0) return '(list)';
+
+  const spaceBefore = formatSpace(space, depth, false);
+  const endSpace = space ? formatSpace(space, depth, true) : '';
+
+  const items = cv.list.map(v => prettyPrintWithDepth(v, space, depth)).join(spaceBefore);
+
+  return `(list${spaceBefore}${items}${endSpace})`;
+}
+
+/**
+ * @description format Tuple clarity values in clarity style strings
+ * with the ability to prettify the result with line break end space indentation
+ * @exemple
+ * ```ts
+ * formatTuple(Cl.tuple({ id: Cl.uint(1) }))
+ * // { id: u1 }
+ *
+ * formatTuple(Cl.tuple({ id: Cl.uint(1) }, 2))
+ * // {
+ * //   id: u1
+ * // }
+ * ```
+ */
+function formatTuple(cv: TupleCV, space: number, depth = 1): string {
+  if (Object.keys(cv.data).length === 0) return '{}';
+
+  const items: string[] = [];
+  for (const [key, value] of Object.entries(cv.data)) {
+    items.push(`${key}: ${prettyPrintWithDepth(value, space, depth)}`);
+  }
+
+  const spaceBefore = formatSpace(space, depth, false);
+  const endSpace = formatSpace(space, depth, true);
+
+  return `{${spaceBefore}${items.join(`,${spaceBefore}`)}${endSpace}}`;
+}
+
+function exhaustiveCheck(param: never): never {
+  throw new Error(`invalid clarity value type: ${param}`);
+}
+
+// the exported function should not expose the `depth` argument
+function prettyPrintWithDepth(cv: ClarityValue, space = 0, depth: number): string {
+  if (cv.type === ClarityType.BoolFalse) return 'false';
+  if (cv.type === ClarityType.BoolTrue) return 'true';
+
+  if (cv.type === ClarityType.Int) return cv.value.toString();
+  if (cv.type === ClarityType.UInt) return `u${cv.value.toString()}`;
+
+  if (cv.type === ClarityType.StringASCII) return `"${cv.data}"`;
+  if (cv.type === ClarityType.StringUTF8) return `u"${cv.data}"`;
+
+  if (cv.type === ClarityType.PrincipalContract) return `'${principalToString(cv)}`;
+  if (cv.type === ClarityType.PrincipalStandard) return `'${principalToString(cv)}`;
+
+  if (cv.type === ClarityType.Buffer) return formatBuffer(cv);
+
+  if (cv.type === ClarityType.OptionalNone) return 'none';
+  if (cv.type === ClarityType.OptionalSome)
+    return `(some ${prettyPrintWithDepth(cv.value, space, depth)})`;
+
+  if (cv.type === ClarityType.ResponseOk)
+    return `(ok ${prettyPrintWithDepth(cv.value, space, depth)})`;
+  if (cv.type === ClarityType.ResponseErr)
+    return `(err ${prettyPrintWithDepth(cv.value, space, depth)})`;
+
+  if (cv.type === ClarityType.List) {
+    return formatList(cv, space, depth + 1);
+  }
+  if (cv.type === ClarityType.Tuple) {
+    return formatTuple(cv, space, depth + 1);
+  }
+
+  // make sure that we exhausted all ClarityTypes
+  exhaustiveCheck(cv);
```

Comment:
Interesting pattern 👍 I like the exhaustive check. Is that why you didn't go for a switch?

Reply:
I generally don't use `switch` often because I find it has an odd syntax within the JS ecosystem.
But in TS it has the advantage of checking exhaustiveness. But the exhaustiveCheck can be useful to make TS happy sometimes.

If it fits within Stacks.js coding style, I can change for `switch`, let me know what you prefer. In this particular code sample it would make sense I guessx

Reply:
Either is fine with me, 🚀